src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -152,13 +154,131 @@ using Parser = std::function<O(const Node&, DefaultRandom&)>;
+                   DefaultRandom& rng,
+                   const std::map<std::string, Parser<Out>>& parsers);
 
 template <bool Verbatim>
 std::unique_ptr<DocumentGenerator::Impl> documentGenerator(const Node& node, DefaultRandom& rng);
 
 template <bool Verbatim>
 UniqueGenerator<bsoncxx::array::value> arrayGenerator(const Node& node, DefaultRandom& rng);
 
+template <typename distribution,
+          const char* diststring,
+          const char* param1name,
+          const char* param2name>
+class DoubleGenerator2Parm : public Generator<double> {
+public:
+    DoubleGenerator2Parm(const Node& node, DefaultRandom& rng)
+        : _rng{rng},
+          _param1Gen{doubleGenerator(extract(node, param1name, diststring), _rng)},
+          _param2Gen{doubleGenerator(extract(node, param2name, diststring), _rng)} {}
+
+    double evaluate() override {
+        auto param1 = _param1Gen->evaluate();
+        auto param2 = _param2Gen->evaluate();
+        auto dist = distribution{param1, param2};
+        return dist(_rng);
+    }
+
+private:
+    DefaultRandom& _rng;
+    UniqueGenerator<double> _param1Gen;
+    UniqueGenerator<double> _param2Gen;
+};
+
+template <typename distribution, const char* diststring, const char* param1name>
+class DoubleGenerator1Parm : public Generator<double> {
+public:
+    DoubleGenerator1Parm(const Node& node, DefaultRandom& rng)
+        : _rng{rng}, _param1Gen{doubleGenerator(extract(node, param1name, diststring), _rng)} {}
+
+    double evaluate() override {
+        auto param1 = _param1Gen->evaluate();
+        auto dist = distribution{param1};
+        return dist(_rng);
+    }
+
+private:
+    DefaultRandom& _rng;
+    UniqueGenerator<double> _param1Gen;
+};
+
+// Constant strings for arguments for templates
+static const char astr[] = "a";
+static const char bstr[] = "b";
+static const char kstr[] = "k";
+static const char mstr[] = "m";
+static const char nstr[] = "n";
+static const char sstr[] = "s";
+static const char minstr[] = "min";
+static const char maxstr[] = "max";
+static const char alphastr[] = "alpha";
+static const char betastr[] = "beta";
+static const char lambdastr[] = "lambda";
+static const char meanstr[] = "mean";
+static const char medianstr[] = "median";
+static const char sigmastr[] = "sigma";
+
+// constant strings for distribution names in templates
+static const char uniformstr[] = "uniform";
+static const char exponentialstr[] = "exponential";
+static const char gammastr[] = "gamma";
+static const char weibullstr[] = "weibull";
+static const char extremestr[] = "extreme_value";
+static const char laplacestr[] = "laplace";
+static const char normalstr[] = "normal";
+static const char lognormalstr[] = "lognormal";
+static const char chisquaredstr[] = "chi_squared";
+static const char noncentralchisquaredstr[] = "non_central_chi_squared";
+static const char cauchystr[] = "cauchy";
+static const char fisherfstr[] = "fisher_f";
+static const char studenttstr[] = "student_t";
```

Comment:
I always forget you can have `const char[]` as template params. This is quite nice and saves a lot of boilerplate.

Reply:
I wish I could pass them in directly and not define them separately, but oh well.